packages/flutter/lib/src/rendering/semantics.dart

```diff
@@ -0,0 +1,456 @@
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui' show Rect;
+
+import 'package:flutter/painting.dart';
+import 'package:sky_services/semantics/semantics.mojom.dart' as engine;
+import 'package:vector_math/vector_math_64.dart';
+
+import 'node.dart';
+
+typedef void SemanticAnnotator(SemanticsNode semantics);
+
+/// Interface for RenderObjects to implement when they want to support
+/// being tapped, etc.
+///
+/// These handlers will only be called if the relevant flag is set
+/// (e.g. handleSemanticTap() will only be called if canBeTapped is
+/// true, handleSemanticScrollDown() will only be called if
+/// canBeScrolledVertically is true, etc).
+abstract class SemanticActionHandler {
+  void handleSemanticTap() { }
+  void handleSemanticLongPress() { }
+  void handleSemanticScrollLeft() { }
+  void handleSemanticScrollRight() { }
+  void handleSemanticScrollUp() { }
+  void handleSemanticScrollDown() { }
+}
+
+enum _SemanticFlags {
+  mergeAllDescendantsIntoThisNode,
+  canBeTapped,
+  canBeLongPressed,
+  canBeScrolledHorizontally,
+  canBeScrolledVertically,
+  hasCheckedState,
+  isChecked,
+}
+
+/// A BitField over an enum (or other class whose values implement "index").
+/// Only the first 63 values of the enum can be used as indices.
+class BitField<T extends dynamic> {
+  static const _kSMIBits = 63; // see https://www.dartlang.org/articles/numeric-computation/#smis-and-mints
+  static const _kAllZeros = 0;
+  static const _kAllOnes = 0x7FFFFFFFFFFFFFFF; // 2^(_kSMIBits+1)-1
+  BitField(this._length) : _bits = _kAllZeros {
+    assert(_length <= _kSMIBits);
+  }
+  BitField.filled(this._length, bool value) : _bits = value ? _kAllOnes : _kAllZeros {
+    assert(_length <= _kSMIBits);
+  }
+  final int _length;
+  int _bits;
+  bool operator [](T index) {
+    assert(index.index < _length);
+    return (_bits & 1 << index.index) > 0;
+  }
+  void operator []=(T index, bool value) {
+    assert(index.index < _length);
+    if (value)
+      _bits = _bits | (1 << index.index);
+    else
+      _bits = _bits & ~(1 << index.index);
+  }
+  void reset([ bool value = false ]) {
+    _bits = value ? _kAllOnes : _kAllZeros;
+  }
+}
+
+typedef bool SemanticsNodeVisitor(SemanticsNode node);
+
+class SemanticsNode extends AbstractNode {
+  SemanticsNode({
+    SemanticActionHandler handler
+  }) : _id = _generateNewId(),
+       _actionHandler = handler;
+
+  SemanticsNode.root({
+    SemanticActionHandler handler
+  }) : _id = 0,
+       _actionHandler = handler {
+    attach();
+  }
+
+  static int _lastIdentifier = 0;
+  static int _generateNewId() {
+    _lastIdentifier += 1;
+    return _lastIdentifier;
+  }
+
+  final int _id;
+  final SemanticActionHandler _actionHandler;
+
+
+  // GEOMETRY
+  // These are automatically handled by RenderObject's own logic
+
+  Matrix4 get transform => _transform;
+  Matrix4 _transform = new Matrix4.identity();
```

Comment:
To avoid churn even further, I've made all the code handle "null" as meaning "identity matrix". This avoids us accidentally mutating our "identity" static.